pkg/kubelet/util/node_startup_latency_tracker.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/utils/clock"
+)
+
+const (
+	kubeletStart    = "kubelet_start"
+	attemptRegister = "attempt_register"
+	registered      = "registered"
+	nodeReady       = "node_ready"
```

Comment:
@dgrisonnet we need the granularity because of the intermediate operations that happen in between, mainly the network initialization that is returned as part of the runtime. Having all these steps is easier for the operator or developer to identify the source or at least the part of the initialization that impacts the startup latency

Reply:
example, cluster with 1000 nodes, some of the nodes will be impacted by
- cni initialization: kubelet CRI runtime status call
- cloud-provider initialization: time the cloud-provider will see the node and remove taints or update fields like PodCIDRs
- apiserver: 1000 request more or less at same time , some of them will be queued

> From a monitoring perspective I find it quite hard to understand what's happening and even harder to start investigating.

kubelet start -> attempt_to_register = kubelet or node contention
attempt_to_register -> registered =. apiserver issues or kubelet self checks
registered > node ready = external actors like CNI, cloud-provider or kubelet checks

> if the overall node startup time is not impacted, knowing that these steps were slow is not very useful.

That is actually the problem, that makes that startup latency node issues got directed always to the same team because there is no way to differentiate which component is causing the problem, this will make easy to triage and open the issue to the right component as it will be clearer which actors are involved on the latency in each stage

> but I think it would be better suited in logs/traces and just keep the overall duration in a metrics

we need to expose these metrics  to build a SLO on a cluster, is a cluster wide metric

Reply:
> That is actually the problem, that makes that startup latency node issues got directed always to the same team because there is no way to differentiate which component is causing the problem, this will make easy to triage and open the issue to the right component as it will be clearer which actors are involved on the latency in each stage

That makes a bit more sense after your explanation.

But I still find it hard to make the following transparent to the users:
```
kubelet start -> attempt_to_register = kubelet or node contention
attempt_to_register -> registered =. apiserver issues or kubelet self checks
registered > node ready = external actors like CNI, cloud-provider or kubelet checks
```

Both `kubelet start -> attempt_to_register` and `registered -> node ready` would be umbrella measurement in the same way as `kubelet start -> node ready` is one, but they are less explicit.

What would you think about keeping `node_startup_duration_seconds` as the only umbrella metric catching all kind of node startup latency issues and then have very specific latency metrics giving a more precise idea of where the problem is? Such as:
- node_startup_registration_duration_seconds
- node_startup_cni_initialization_duration_seconds
- node_startup_cloud_provider_initialization_duration_seconds
- ...

I still don't know how I would transform `kubelet start -> attempt_to_register` in something more comprehensive though

Reply:
I am leaning towards keeping `kubelet start -> attempt_to_register` and `registered -> node ready` mostly because 

a) we probably cannot exhaust all operations during node startup and give each of them a metric, and
b) if in future we add more steps to node startup process, we will have them covered with these umbrella metrics.

That said, I like the idea of adding specific latency metrics . They will help us catch regression when making code changes on those components specifically. Maybe we can add them as a separate PR?

Reply:
That's a fair argument :+1:

I would still suggest splitting into different metrics to make the metrics easier to discover and understand. Maybe something like:
```
node_startup_pre_registration_duration_seconds
node_startup_registration_duration_seconds
node_startup_post_registration_duration_seconds
node_startup_duration_seconds
```
> Maybe we can add them as a separate PR?

Yes definitely. My previous comment was just some suggestions on how to cover the SLO story behind node startup.

Reply:
Thanks guys for the discussion. 

> But I still find it hard to make the following transparent to the users:

> I still don't know how I would transform kubelet start -> attempt_to_register in something more comprehensive though

@dgrisonnet I believe your argument is around describing a metric precisely with one **step name**. That was not my intention. I was trying to describe a metric with **2 milestone names**. I want to focus on milestones, not any specific steps in between. "Milestone metrics" and "step metrics" serve different purpose, like @ruiwen-zhao said.

Reply:
```
node_startup_pre_registration_duration_seconds node_startup_registration_duration_seconds node_startup_post_registration_duration_seconds node_startup_duration_seconds
```

This also looks good to me. Thank you @dgrisonnet for giving them clear names . @pacoxu Are you ok with this?

Reply:
This looks good to me.

Reply:
sg, thank you for confirming. Let me update them.

Reply:
Updated the metrics. I add another one `node_startup_pre_kubelet_duration_seconds` from node boot time to kubelet start time.